### PR TITLE
feat(zipkin): improves zipkin exporter.

### PR DESCRIPTION
**Pull request by jcchavezs**
This PR changes the exporter signature to provide a way to flush the queue. This is specially important in order to get control of the exporter decoupled from the application lifecycle and getting rid from [this kind of hacks](https://github.com/open-telemetry/opentelemetry-go/pull/1319/files#diff-6b482021e206d78ba5adea36217c31fa732aee0872682fcb95c29b6750756078L67).

### CODE REVIEW

File: exporters/trace/zipkin/zipkin.go [hunk 1]
```diff
@@ -166,19 +166,12 @@ func (e *Exporter) ExportSpans(ctx context.Context, batch []*export.SpanData) er
...
 	}
-	e.logf("zipkin responded with status %d", resp.StatusCode)
+	defer resp.Body.Close()
 
-	_, err = ioutil.ReadAll(resp.Body)
-	if err != nil {
-		// Best effort to clean up here.
-		resp.Body.Close()
-		return e.errf("failed to read response body: %v", err)
+	if resp.StatusCode != 202 {
```
Comment (jcchavezs) [inline exporters/trace/zipkin/zipkin.go:171]: This is enough for zipkin (and zipkin clones). We don't need to read the body as it does not return meaningful information more than the status code.

File: exporters/trace/zipkin/zipkin.go [hunk 2]
```diff
@@ -111,30 +110,31 @@ func NewRawExporter(collectorURL, serviceName string, opts ...Option) (*Exporter
 
 // NewExportPipeline sets up a complete export pipeline
 // with the recommended setup for trace provider
-func NewExportPipeline(collectorURL, serviceName string, opts ...Option) (*sdktrace.TracerProvider, error) {
-	exp, err := NewRawExporter(collectorURL, serviceName, opts...)
+func NewExportPipeline(collectorURL, serviceName string, opts ...Option) (*sdktrace.TracerProvider, func(), error) {
```
Comment (MrAlias) [inline exporters/trace/zipkin/zipkin.go:113]: Changing this function signature to return a reference to the BSP `ForceFlush` method does not seem necessary. The `TracerProvider.Shutdown` method will call the BSP shutdown method and flush it.

File: exporters/trace/zipkin/zipkin.go [hunk 3]
```diff
@@ -166,19 +166,12 @@ func (e *Exporter) ExportSpans(ctx context.Context, batch []*export.SpanData) er
 	if err != nil {
 		return e.errf("request to %s failed: %v", e.url, err)
 	}
-	e.logf("zipkin responded with status %d", resp.StatusCode)
+	defer resp.Body.Close()
```
Comment (MrAlias) [inline exporters/trace/zipkin/zipkin.go:169]: The original change to read the whole response body was included so the default HTTP client's Transport will
reuse HTTP/1.x "keep-alive" TCP connections. It seems like a good idea to keep this behavior.

https://github.com/open-telemetry/opentelemetry-go/issues/786#issuecomment-637633843

**Review by MrAlias — CHANGES_REQUESTED** [review]
I think this PR can be pruned down to focus on the meaningful change of checking the response status of the Zipkin server and the minor updates to documentation.

File: exporters/trace/zipkin/zipkin.go (hunk 2 above)
Comment (jcchavezs) [reply exporters/trace/zipkin/zipkin.go:113]: I think user should have a way to flush traces despite the app lifecycle. Local development is a very common use case.

File: exporters/trace/zipkin/zipkin.go (hunk 3 above)
Comment (jcchavezs) [reply exporters/trace/zipkin/zipkin.go:169]: Even if that is the case, the main feedback should come from http status code. If you want to, we can read the body (and discard it) in a defer. Does that make sense?

Comment (MrAlias) [reply exporters/trace/zipkin/zipkin.go:169]: Verification of the status code is a good addition :+1:. I would prefer to preserve the existing body handling and error logging.

File: exporters/trace/zipkin/zipkin.go (hunk 2 above)
Comment (MrAlias) [reply exporters/trace/zipkin/zipkin.go:113]: These functions are provided as a convince for users that do not want to have to deal with details of things like the BSP. If a user would like to have more control over the flushing of the BSP used, they should forego using these convenience functions and directly construct the trace export pipeline in a way that suits their manner.

Comment (jcchavezs) [reply exporters/trace/zipkin/zipkin.go:113]: I find it a bit unfair as for example jaeger reporter provides this functionality.

Comment (MrAlias) [reply exporters/trace/zipkin/zipkin.go:113]: > I find it a bit unfair as for example jaeger reporter provides this functionality.

Oh, sorry, I didn't mean to imply this was a subjective choice to limit the Zipkin exporter only. That is a good point, we should update the Jaeger exporter package to match this. The addition of having the `TracerProvider.Shutdown` method stop all the BSPs is somewhat recent and I imagine we missed this fix when we rolled it out.

Comment (MrAlias) [reply exporters/trace/zipkin/zipkin.go:113]: Tracking the Jaeger change here: https://github.com/open-telemetry/opentelemetry-go/issues/1327

**Pull request closed without merging**